File: src/custom_vtol_plugin/scripts/custom_bridge.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray

from gz.transport import Node as GzNode
from gz.msgs.double_v_pb2 import Double_V

class CustomBridge(Node):
    def __init__(self):
        super().__init__('custom_bridge')
        self.sub = self.create_subscription(Float64MultiArray,
                                            '/ros_motor_command',
                                            self.callback,
                                            10)
        self.gz_node = GzNode()
        self.pub = self.gz_node.advertise('/model/standard_vtol/command/motor_speed', Double_V)

    def callback(self, msg):
        gz_msg = Double_V()
        gz_msg.data.extend(msg.data)
        self.pub.publish(gz_msg)
        self.get_logger().info(f"Bridged {msg.data} → Gazebo Double_V")

def main():
    rclpy.init()
    node = CustomBridge()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/custom_bridge_cpp/src/custom_bridge.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float64_multi_array.hpp>

#include <gz/transport/Node.hh>
#include <gz/msgs/double_v.pb.h>

class CustomBridge : public rclcpp::Node {
public:
  CustomBridge() : Node("custom_bridge") {
    // Declare parameter with default topic
    this->declare_parameter<std::string>(
      "motor_topic", "/model/standard_vtol/command/motor_speed");

    // Get parameter value
    std::string topic = this->get_parameter("motor_topic").as_string();

    
    pub_ = gz_node_.Advertise<gz::msgs::Double_V>(topic);

    // Subscribe to ROS motor command
    sub_ = this->create_subscription<std_msgs::msg::Float64MultiArray>(
      "/ros_motor_command", 10,
      std::bind(&CustomBridge::callback, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "Publishing to Gazebo topic: %s", topic.c_str());
  }

private:
  void callback(const std_msgs::msg::Float64MultiArray::SharedPtr msg) {
    gz::msgs::Double_V gz_msg;
    for (auto val : msg->data) {
      gz_msg.add_data(val);
    }
    pub_.Publish(gz_msg);
    RCLCPP_INFO(this->get_logger(), "Bridged %zu values to Gazebo", msg->data.size());
  }

  gz::transport::Node gz_node_;
  gz::transport::Node::Publisher pub_;
  rclcpp::Subscription<std_msgs::msg::Float64MultiArray>::SharedPtr sub_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<CustomBridge>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
